src/AppInstallerCommonCore/Registry.cpp

```diff
@@ -93,6 +177,87 @@ namespace AppInstaller::Registry
         return m_type == type;
     }
 
+    ValueList::ValueRef::ValueRef(std::wstring&& valueName, DWORD type, std::vector<BYTE>&& data) : Value(type, std::move(data)), m_valueName(std::move(valueName)) {}
+
+    std::string ValueList::ValueRef::Name() const
+    {
+        return Utility::ConvertToUTF8(m_valueName);
+    }
+
+    ValueList::const_iterator& ValueList::const_iterator::operator++()
+    {
+        ++m_index;
+        GetValue();
+        return *this;
+    }
+
+    ValueList::const_iterator ValueList::const_iterator::operator++(int)
+    {
+        const_iterator result;
+        result.m_key = m_key;
+        result.m_index = m_index++;
+        result.m_value = std::nullopt;
+        std::swap(m_value, result.m_value);
+        GetValue();
+        return result;
+    }
+
+    bool ValueList::const_iterator::operator==(const const_iterator& other) const
+    {
+        return (!m_key && !other.m_key) || (m_key.get() == other.m_key.get() && m_index == other.m_index);
+    }
+
+    bool ValueList::const_iterator::operator!=(const const_iterator& other) const
+    {
+        return !operator==(other);
+    }
+
+    void ValueList::const_iterator::GetValue()
+    {
+        std::wstring valueName;
+        if (!TryGetRegistryValueNameFromIndex(m_key, m_index, valueName))
+        {
+            m_key.reset();
+            return;
+        }
+
+        DWORD type;
+        std::vector<BYTE> data;
+        if (!TryGetRegistryValueData(m_key, valueName, type, data))
+        {
+            THROW_HR(E_UNEXPECTED);
+        }
+
+        m_value = ValueRef{ std::move(valueName), type, std::move(data) };
+    }
+
+    const ValueList::ValueRef& ValueList::const_iterator::operator*() const
+    {
+        return m_value.value();
+    }

```

Comment:
Not that it matters much for the general use case, but I would err toward making the caller keep the value while they wanted it rather than the iterator getting the value while iterating.  If I wanted to skip ahead for some reason this way reads every value anyway.